### PR TITLE
[Core] Adding new constructor to ReadMaterialUtility

### PR DESCRIPTION
**Pull request by loumalouomega**
Separately from #2919

### CODE REVIEW

**Comment by loumalouomega** [comment]
As I said in #2919 the current constructor is very very very rigid and requires the existence  of a "physical" json file to read properties. With this constructor is is possible to read properties from a Parameter object (both from python and cpp). 

This is completely necessary to create cpp tests, otherwise you need to create the json file with all the problems associated of reading a file from c++. I can remove the interface on python, but I think not allowing to do this is very troublesome for standalone test creation

**Comment by loumalouomega** [comment]
As I said I am using in in #2933:

https://github.com/KratosMultiphysics/Kratos/blob/e674e4e469be2f626bcf80cf53e96ffb1bcaa769/applications/StructuralMechanicsApplication/tests/cpp_tests/test_rule_of_mixtures_law.cpp#L179

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
blocking while discussing this

**Comment by loumalouomega** [comment]
I already gave you my arguments

**Comment by philbucher** [comment]
I am fine with this, it makes sense from my side

**Comment by frastellini** [comment]
Initially my vote is YES. But it would be good to listen the arguments against.

**Comment by RiccardoRossi** [comment]
The point is that i really do not understand why you need "Model" to read a file.

Giving the Model as parameter gives access to everything, so is the worst possible thing to do from the point of view of data hiding. Why do you want to do that?

you could simply open a file and read from it. There is nothing difficult in doing it (neither from python nor from c++, at least as long as you use ascii characters in the name, which we can definitely ensure at least for the tests.

**Comment by philbucher** [comment]
@RiccardoRossi this utility also assings the read materials/properties to the elements/conditions, which are stored in the ModelPart/Model

**Comment by loumalouomega** [comment]
@RiccardoRossi this is the utility used to assign the properties to the whole model. See current master constructor (which stays with this PR): 

https://github.com/KratosMultiphysics/Kratos/blob/171de4886562dafdc6c36dd8379fae77cd2c58a3/kratos/utilities/read_materials_utility.cpp#L24

It uses a Model and a Parameter, but this Parameter **only** contains the location of the materials file.

**Comment by RiccardoRossi** [comment]
Until now we passed modelpart which is (a tiny little) more restrictive than model. Why do we need to upgrade that to model? (i mean, the different is somewhat aesthetical, since once passed a modelpart you can get the owner_model, however conceptually i do not think it is the correct direction)

aside of this, why we do not use directly the filename in here:

     ReadMaterialsUtility::ReadMaterialsUtility(
        const std::string& rParametersName,
        Model& rModel
        ) : mrModel(rModel)
    {
         // Receive json string with materials properties, create Parameters
         Parameters materials(rParametersName);
     
         GetPropertyBlock(materials);
     }

**Comment by philbucher** [comment]
@RiccardoRossi the Model, not the ModelPart is being passed (since I started which was a while ago :) )
[this is Master](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/read_materials_utility.h#L78-L80)

your comment also makes sense ...

**Comment by loumalouomega** [comment]
> @RiccardoRossi this is the utility used to assign the properties to the whole model. See current master constructor (which stays with this PR):
> 
> [Kratos/kratos/utilities/read_materials_utility.cpp](https://github.com/KratosMultiphysics/Kratos/blob/171de4886562dafdc6c36dd8379fae77cd2c58a3/kratos/utilities/read_materials_utility.cpp#L24)
> 
> Line 24 in [171de48](/KratosMultiphysics/Kratos/commit/171de4886562dafdc6c36dd8379fae77cd2c58a3)
>  ReadMaterialsUtility::ReadMaterialsUtility( 
> 
> It uses a Model and a Parameter, but this Parameter **only** contains the location of the materials file.

If you check the link I put in here (which is the file in the master), you will see we are using the Model (and I explicitly wrote that in my reply)

**Comment by loumalouomega** [comment]
> @RiccardoRossi the Model, not the ModelPart is being passed (since I started which was a while ago :) )
> [this is Master](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/read_materials_utility.h#L78-L80)
> 
> your comment also makes sense ...

I basically quote the same constructor but in the cpp

**Comment by philbucher** [comment]
@RiccardoRossi I digged a bit through the history, it was you that did the very first version of this utility (in python)
This gets THE MODEL and not the modelpart => https://github.com/KratosMultiphysics/Kratos/commit/15b42598347656cd55b57ada29cbd06e871f7d7d

@loumalouomega nevertheless he is right that you can do what you want to do just by passing the string

**Comment by RiccardoRossi** [comment]
oh ... then i am the one to blame? well... i still really do not like it. 
However...if it is already like this in the master i will accept the change... but i still think that we should use the filename directly and not through a parameter

**Comment by loumalouomega** [comment]
But then @RiccardoRossi you are forcing to create the cpp test and read a file, and I promise you, because I tried reading directly from cpp is a pain in the ass because of the OS problem (the test I created didn't run on Windows), that's why I added the constructor. It was a last resource after trying other options with the current implementation

**Comment by RiccardoRossi** [comment]
@loumalouomega i really do not understand you. Why adding the Model solves your problem with reading? Anyhow, if this file is to be used from c++, than you **need** to be able to open the file correctly...

**Comment by loumalouomega** [comment]
@RiccardoRossi see the code of the current constructor:

~~~cpp
ReadMaterialsUtility::ReadMaterialsUtility(
    Parameters Params,
    Model& rModel
    ) : mrModel(rModel)
{
    Parameters default_parameters(R"(
    {
        "Parameters" : {
            "materials_filename" : "please specify the file to be opened"
        }
    }  )"
    );

    Params.RecursivelyValidateAndAssignDefaults(default_parameters);

    // Read json string in materials file, create Parameters
    const std::string& materials_filename = Params["Parameters"]["materials_filename"].GetString();
    std::ifstream infile(materials_filename);
    KRATOS_ERROR_IF_NOT(infile.good()) << "Materials file: " << materials_filename << " cannot be found" << std::endl;
    std::stringstream buffer;
    buffer << infile.rdbuf();
    Parameters materials(buffer.str());

    GetPropertyBlock(materials);
}
~~~

You pass Model + Parameters but this parameters **only** contains the location of the json file where the materials are defined. What happens if I want to read properties from object Parameters? 

You can't. So this constructor works in two steps:

- Creates the object ReadMaterialsUtility.
- You use the method ReadMaterials to read the properties directly from a Parameters object.

This can be seen in the example I linked previously:

~~~cpp
    // Read properties
    auto read_util = ReadMaterialsUtility(rThisModelPart.GetModel());
    read_util.ReadMaterials(parameters);
~~~

File: kratos/utilities/read_materials_utility.cpp [hunk 1]
```diff
@@ -64,6 +71,14 @@ ReadMaterialsUtility::ReadMaterialsUtility(
 /***********************************************************************************/
 /***********************************************************************************/
 
+void ReadMaterialsUtility::ReadMaterials(Parameters Params)
```
Comment (RiccardoRossi) [inline kratos/utilities/read_materials_utility.cpp]: just see the comment below (in the header)

File: kratos/utilities/read_materials_utility.h [hunk 2]
```diff
@@ -75,19 +78,43 @@ class KRATOS_API(KRATOS_CORE) ReadMaterialsUtility
...
-
     ///@}
     ///@name Operations
     ///@{
 
+    /**
+     * @brief This reads the properties from parameters
+     * @param Params The configuration parameters defining the properties
+     */
+    void ReadMaterials(Parameters Params);
```
Comment (RiccardoRossi) [inline kratos/utilities/read_materials_utility.h]: maybe give a significative name.

something like "MaterialData"

      void ReadMaterials(Parameters MaterialData);

**Comment by loumalouomega** [comment]
Comments adressed @RiccardoRossi

File: kratos/utilities/read_materials_utility.cpp [hunk 3]
```diff
@@ -64,6 +71,14 @@ ReadMaterialsUtility::ReadMaterialsUtility(
 /***********************************************************************************/
 /***********************************************************************************/
 
+void ReadMaterialsUtility::ReadMaterials(Parameters MaterialData)
```
Comment (RiccardoRossi) [inline kratos/utilities/read_materials_utility.cpp:74]: wouldn't it be better to remove overall the constructor 

      ReadMaterialsUtility( Params, model )

and to have two methods 

      ReadMaterials(std::string filename )
 
and 

      ReadMaterials(Parameters MaterialData)

**Review by RiccardoRossi — APPROVED** [review]
I am approving, however the solution i am putting in my last comment would be nicer

File: kratos/utilities/read_materials_utility.cpp (hunk 3 above)
Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp:74]: Then we will remove the current method breaking the retrocompatibility

**Pull request merged**